core/teams/engine.go

```diff
@@ -0,0 +1,145 @@
+package teams
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"code.vegaprotocol.io/vega/core/types"
+	proto "code.vegaprotocol.io/vega/protos/vega"
+)
+
+var (
+	ErrTeamLeaderCannotJoinAnotherTeam = errors.New("a team leader cannot join another team")
+)
+
+//go:generate go run github.com/golang/mock/mockgen -destination mocks/mocks.go -package mocks code.vegaprotocol.io/vega/core/teams EpochEngine
+
+type EpochEngine interface {
+	NotifyOnEpoch(func(context.Context, types.Epoch), func(context.Context, types.Epoch))
+}
+
+type Engine struct {
+	// teams tracks all teams by team ID.
+	teams map[TeamID]*Team
+
+	// allTeamMembers tracks all the parties that belongs to a team, leaders and
+	// follower, by their current team ID.
+	allTeamMembers map[Party]TeamID
+
+	// membersToMove tracks all the parties that switch teams. The switch only
+	// happens by the end of the epoch.
+	membersToMove map[Party]teamMove
+}
+
+func (e *Engine) CreateTeam(leader Party, name string, teamURL string, avatarURL string, enableRewards bool) (TeamID, error) {
+	newTeamID := e.newUniqueTeamID()
+
+	if _, alreadyMember := e.allTeamMembers[leader]; alreadyMember {
+		return "", fmt.Errorf("the party %q already belongs to a team", leader)
+	}
+
+	e.teams[newTeamID] = &Team{
+		ID:            newTeamID,
+		Leader:        leader,
+		Name:          name,
+		TeamURL:       teamURL,
+		AvatarURL:     avatarURL,
+		EnableRewards: enableRewards,
+	}
+
+	e.allTeamMembers[leader] = newTeamID
+
+	return newTeamID, nil
+}
+
+func (e *Engine) UpdateTeam(id TeamID, name string, teamURL string, avatarURL string, enableRewards bool) error {
+	_, exists := e.teams[id]
+	if !exists {
+		return errNoTeamMatchesID(id)
+	}
+
+	e.teams[id].Name = name
+	e.teams[id].TeamURL = teamURL
+	e.teams[id].AvatarURL = avatarURL
+	e.teams[id].EnableRewards = enableRewards
+
+	return nil
+}
+
+func (e *Engine) JoinTeam(id TeamID, party Party) error {
```

Comment:
Should we implement the events being sent when a new team is joined, or created, etc?

Reply:
This is another issue.

Reply:
fair enough.